services/brig/test/unit/Test/Brig/Calling.hs

```diff
@@ -1,241 +0,0 @@
-{-# LANGUAGE RecordWildCards #-}
-
--- This file is part of the Wire Server implementation.
---
--- Copyright (C) 2020 Wire Swiss GmbH <opensource@wire.com>
---
--- This program is free software: you can redistribute it and/or modify it under
--- the terms of the GNU Affero General Public License as published by the Free
--- Software Foundation, either version 3 of the License, or (at your option) any
--- later version.
---
--- This program is distributed in the hope that it will be useful, but WITHOUT
--- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
--- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
--- details.
---
--- You should have received a copy of the GNU Affero General Public License along
--- with this program. If not, see <https://www.gnu.org/licenses/>.
-
-module Test.Brig.Calling where
-
-import Brig.Calling
-import Brig.Options
-import Brig.PolyLog
-import Control.Retry
-import Data.List.NonEmpty (NonEmpty (..))
-import qualified Data.List.NonEmpty as NonEmpty
-import Data.Range
-import qualified Data.Set as Set
-import Imports
-import Network.DNS
-import Polysemy
-import qualified System.Logger as Log
-import Test.Tasty
-import Test.Tasty.HUnit
-import qualified UnliftIO.Async as Async
-import Wire.Network.DNS.Effect
-import Wire.Network.DNS.SRV
-
-data FakeDNSEnv = FakeDNSEnv
-  { fakeLookupFn :: Domain -> SrvResponse,
-    fakeLookupCalls :: IORef [Domain]
-  }
-
-newFakeDNSEnv :: (Domain -> SrvResponse) -> IO FakeDNSEnv
-newFakeDNSEnv lookupFn =
-  FakeDNSEnv lookupFn <$> newIORef []
-
-runFakeDNSLookup :: Member (Embed IO) r => FakeDNSEnv -> Sem (DNSLookup ': r) a -> Sem r a
-runFakeDNSLookup FakeDNSEnv {..} = interpret $ \case
-  LookupSRV domain -> do
-    modifyIORef' fakeLookupCalls (++ [domain])
-    pure $ fakeLookupFn domain
-
-newtype LogRecorder = LogRecorder {recordedLogs :: IORef [(Log.Level, LByteString)]}
-
-newLogRecorder :: IO LogRecorder
-newLogRecorder = LogRecorder <$> newIORef []
-
-recordLogs :: Member (Embed IO) r => LogRecorder -> Sem (PolyLog ': r) a -> Sem r a
-recordLogs LogRecorder {..} = interpret $ \(PolyLog lvl msg) ->
-  modifyIORef' recordedLogs (++ [(lvl, Log.render (Log.renderDefault ", ") msg)])
-
-ignoreLogs :: Sem (PolyLog ': r) a -> Sem r a
-ignoreLogs = interpret $ \(PolyLog _ _) -> pure ()
-
-{-# ANN tests ("HLint: ignore" :: String) #-}
-tests :: TestTree
-tests =
-  testGroup "Calling" $
```

Comment:
While it's a good catch to notice that the DNS library can be used differently (and thus simplify some code), it's not a reason to delete most of the SFT tests without replacing them.

We have some behaviour that should have a test in order for it not to disappear in the next refactoring:
- using a maximum of only X SFT servers even if there are more
- randomization per-request (even if the list of SFT in SRV stays the same) for clients
- fail-safe behaviour in times where the SRV lookup fails temporarily (as has happened a few times in the past) to still give out the current list of servers
- no SFT in config when some config variable is not set

Can you please make sure to cover these cases with a test in this PR. Perhaps you need an integration test here now.

Reply:
Yep I'll re-add these tests. I agree they're useful to keep.

> fail-safe behaviour in times where the SRV lookup fails temporarily (as has happened a few times in the past) to still give out the current list of servers

@akshaymankar brought this up too.  But I'm not sure breaking cache invalidation in the presence of failures is the best solution here because it's hard to distinguish between true and transient failures. I think the change of behavior is a good one but I'm happy to be convinced otherwise.

 Instead we should tweak [`resolvRetry`](https://hackage.haskell.org/package/dns-4.0.1/docs/Network-DNS-Resolver.html#v:resolvRetry) to better survive transient failures and we should tweak the record's TTL for more  "permanent"  transient failures.  Currently the SRV records have a TTL of 10 seconds. We could increase that to a few minutes instead.  If there is a positive hit the `dns` library will not contact the DNS server again in that TTL time period.

>  (as has happened a few times in the past) 

Do we have these incidents recorded somewhere?  My gut feeling is that tweaking the TTLs is sufficient to avoid these kind of issues  but perhaps I'm wrong. Knowing exactly the problems we encountered would be useful.

Reply:
Is it this test?

https://github.com/wireapp/wire-server/pull/1271/files#diff-100069ac03fa9aae9c8e7e80a089c378c81ace107d2717d60d77c76cb0f04e4bL135-L149

I am now even more confused about the current behavior.  I  would understand that the record is kept if `SrvResponseError` is returned (A transient faiure) 

but on `SrvNotAvailable` I would expect the record to be properly removed. If the DNS server is giving me an authoritive answer that the record is not available we should not ignore that. How do we ever remove SFT servers from the pool otherwise?

I would not consider `SrvNotAvailable` an "unsuccessful" answer. It is a successful answer; just one that tells there are no entries.

Reply:
Ah this only happens in the scenario where there are _no_ records in the response.

https://github.com/wireapp/wire-server/blob/develop/libs/dns-util/src/Wire/Network/DNS/SRV.hs#L90

If there are no records in the response we keep the previously known set.

Why do we do this though? What scenario can be avoided? Why would a DNS server suddenly stop resolving the entries for the `SRV` record?

Reply:
Regarding `srv-announcer`'s behaviour: if we were the only record left and to be removed, a `ChangeActionUpsert`, with an empty `ResourceRecordSet` would be sent. I don't know how R53 will behave on that. Assuming it will become an empty set, If there's an empty set in DNS, I'd expect brig to not return any records.

If we were to turn off all of SFT, this should be possible by stopping all announcers.

I don't think "sticking with the old view" when no records are returned is correct behaviour here - if the DNS server authoritatively replies there's no records we should honor that.

On the other side, if we get a SERVFAIL, we can only retain the records for the lifetime of the TTL, and after that, we just don't know. But safeguarding against R53 outages is probably out of scope for this.

In all cases, this is all only state still running brigs know, and won't be retained if they're upscaled. We rely on these internal caches as less as possible.